clients/drcachesim/tracer/raw2trace.cpp

```diff
@@ -257,21 +257,40 @@ raw2trace_t::write_syscall_template(raw2trace_thread_data_t *tdata, byte *&buf_i
     buf = buf_base;
     buf_in = reinterpret_cast<byte *>(buf_base); // Defensive: update the returned buf.
 
+    trace_entry_t *buf_last_branch_target_marker = nullptr;
     app_pc saved_decode_pc;
     int inserted_instr_count = 0;
     // XXX i#6495: For now we write out the template as-is to the output trace. But we can
     // potentially customize some properties of the trace. E.g. the start address for the
     // kernel code section.
-    for (const auto &entry : syscall_trace_templates_[syscall_num]) {
+    for (const auto &entry : syscall_trace_templates_[syscall_num].entries) {
         if (type_is_instr(static_cast<trace_type_t>(entry.type)) ||
             // We want to write out at each repstr instance so that we do not accumulate
             // too many buffered entries.
             entry.type == TRACE_TYPE_INSTR_NO_FETCH) {
+            // If this is the last instruction and an indirect branch, and there was a
+            // branch target marker just before it, set it to the fallthrough pc of
+            // the syscall instruction for which we're injecting the trace. This is
+            // simpler than trying to get the actual post-syscall instruction for which
+            // we would perhaps even need to read-ahead to the next raw trace buffer.
+            // XXX i#6495, i#7157: The above strategy does not work for syscalls that
+            // transfer control (like sigreturn), but we do not trace those anyway today
+            // (neither using Intel-PT, nor QEMU) as there are challenges in determining
+            // the post-syscall resumption point.
+            if (type_is_instr_branch(static_cast<trace_type_t>(entry.type)) &&
+                !type_is_instr_direct_branch(static_cast<trace_type_t>(entry.type)) &&
+                inserted_instr_count ==
+                    syscall_trace_templates_[syscall_num].instr_count - 1 &&
+                buf_last_branch_target_marker != nullptr) {
+                buf_last_branch_target_marker->addr =
+                    reinterpret_cast<addr_t>(get_last_pc_fallthrough_if_syscall(tdata));
```

Comment:
Do we need raw2trace unit tests for this, or bust_syscall_inject covers it?

Reply:
burst_syscall_inject covers it. But makes sense to have a raw2trace unit test too.

I added one. But it needed some refactoring, which I split out in #7462.

Reply:
Done.